File: packages/cli/src/report.ts
export const SCHEMA_VERSION = '1.0.0';
export const RUNTIME_VERSION = '1.1.0';

export type Profile = 'cvi' | 'switch' | 'all';
export type Classification = 'normative' | 'advisory' | 'supplemental';
export type Certainty = 'detected' | 'potential' | 'manual';
export type Severity = 'error' | 'warning' | 'info';
export type EvidenceLevel = 'source' | 'computed' | 'runtime' | 'human';

export interface SourceLocation {
  line: number;
  column: number;
  excerpt?: string;
}

export interface ReportItem {
  ruleId: string;
  classification: Classification;
  certainty: Certainty;
  severity: Severity;
  evidence: EvidenceLevel;
  message: string;
  wcag: string[];
  location: SourceLocation | null;
  remediation: string;
  verification: string;
}

export interface AccessibilityReport {
  schemaVersion: string;
  producer: {
    name: 'accessibility-devkit';
    version: string;
    runtime: 'node' | 'python';
  };
  target: {
    kind: 'file' | 'stdin' | 'text' | 'colors' | 'policy';
    value: string;
  };
  summary: {
    errors: number;
    warnings: number;
    info: number;
    findings: number;
    manualChecks: number;
  };
  findings: ReportItem[];
  manualChecks: ReportItem[];
}

export function summarize(
  findings: ReportItem[],
  manualChecks: ReportItem[],
): AccessibilityReport['summary'] {
  return {
    errors: findings.filter((finding) => finding.severity === 'error').length,
    warnings: findings.filter((finding) => finding.severity === 'warning').length,
    info: findings.filter((finding) => finding.severity === 'info').length,
    findings: findings.length,
    manualChecks: manualChecks.length,
  };
}

File: python/src/accessibility_devkit/report.py
"""The versioned runtime-neutral report envelope."""

SCHEMA_VERSION = "1.0.0"
RUNTIME_VERSION = "1.1.0"


def item(rule_id, message, **overrides):
    value = {
        "ruleId": rule_id,
        "classification": "normative",
        "certainty": "detected",
        "severity": "error",
        "evidence": "source",
        "message": message,
        "wcag": [],
        "location": None,
        "remediation": "Review and correct the reported accessibility concern.",
        "verification": "Repeat the check and complete the relevant human or browser verification.",
    }
    value.update(overrides)
    return value


def summarize(findings, manual_checks):
    return {
        "errors": sum(finding["severity"] == "error" for finding in findings),
        "warnings": sum(finding["severity"] == "warning" for finding in findings),
        "info": sum(finding["severity"] == "info" for finding in findings),
        "findings": len(findings),
        "manualChecks": len(manual_checks),
    }


def create_report(kind, target, findings, manual_checks=None, runtime="python"):
    manual_checks = manual_checks or []
    return {
        "schemaVersion": SCHEMA_VERSION,
        "producer": {"name": "accessibility-devkit", "version": RUNTIME_VERSION, "runtime": runtime},
        "target": {"kind": kind, "value": target},
        "summary": summarize(findings, manual_checks),
        "findings": findings,
        "manualChecks": manual_checks,
    }
